### PR TITLE
Adding trace for serviceability

### PR DESCRIPTION
**Pull request by una-tapa**
I started out this PR to add more trace data for preInvoke() method.  
I ended up masking BasicAuth data, which is reported in  #5082 as release bug.  (That's good!)

### CODE REVIEW

**Comment by una-tapa** [comment]
#build

**Comment by una-tapa** [comment]
#build

**Comment by una-tapa** [comment]
#build

**Comment by una-tapa** [comment]
Manual test result 


```
[10/23/19, 20:40:54:591 EDT] 0000003d id=00000000 com.ibm.ws.webcontainer.srt.SRTServletRequest                1 getHeader this->com.ibm.ws.webcontainer31.srt.SRTServletRequest31@dffa8db3:  name --> Authorization header --> Basic ****************
```

```
[10/23/19, 20:43:15:196 EDT] 00000117 id=00000000 com.ibm.ws.webcontainer.srt.SRTServletRequest                1 getHeaderDirect this->com.ibm.ws.webcontainer31.srt.SRTServletRequest31@c3567155:  name --> Authorization header --> Basic ****************
```

```
[10/23/19, 20:43:15:196 EDT] 00000117 id=00000000 .ibm.ws.webcontainer.security.WebAppSecurityCollaboratorImpl 3 Http Header names and values:
Host=[localhost:9080]
User-Agent=[Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:70.0) Gecko/20100101 Firefox/70.0]
Accept=[text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8]
Accept-Language=[en-US,en;q=0.5]
Accept-Encoding=[gzip, deflate]
Connection=[keep-alive]
Upgrade-Insecure-Requests=[1]
Authorization=[Basic xxxxx]
Request Context Path=/basicauth, Servlet Path=/servlet```

**Comment by una-tapa** [comment]
Also made sure WWW-Authenticate header is not masked.  Using the latest build. 

`[10/24/19 19:55:49:818 EDT] 000000ac id=00000000 com.ibm.ws.webcontainer.srt.SRTServletResponse
   1 setHeader  name --> WWW-Authenticate value --> Basic realm="defaultRealm" 
`

**Comment by una-tapa** [comment]
This PR is originally for RTC 268816,  

It will also address  
https://github.com/OpenLiberty/open-liberty/issues/5082

File: dev/com.ibm.ws.transport.http/src/com/ibm/ws/genericbnf/PasswordNullifier.java [hunk 1]
```diff
@@ -19,25 +19,39 @@
     /** Search patterns to use for passwords */
     private static final String PASSWORD_PATTERN = "password=";
     private static final String CLIENT_SECRET_PATTERN = "client_secret=";
+    private static final String BASICAUTH_PATTERN = "basic ";
+    private static final String AUTHORIZATION_HEADER = "Authorization";
```
Comment (wtlucy) [inline dev/com.ibm.ws.transport.http/src/com/ibm/ws/genericbnf/PasswordNullifier.java]: We might want to use `com.ibm.wsspi.http.channel.values.HttpHeaderKeys.HDR_AUTHORIZATION` here instead

File: dev/com.ibm.ws.transport.http/src/com/ibm/ws/genericbnf/PasswordNullifier.java [hunk 2]
```diff
@@ -59,6 +72,36 @@ public static String nullify(String value, byte delimiter) {
...
+        // check to see if we need to null out passwords
+        if (null == value) {
+            return null;
+        }
+
+        String source = value.toLowerCase();
+        StringBuilder b = new StringBuilder(value);
+        boolean modified = optionallyMaskChars(b, source, delimiter, PASSWORD_PATTERN);
+        modified = optionallyMaskChars(b, source, delimiter, CLIENT_SECRET_PATTERN) || modified;
+        if (header.equalsIgnoreCase(com.ibm.wsspi.http.channel.values.HttpHeaderKeys.HDR_AUTHORIZATION.getName())) {
```
Comment (una-tapa) [inline dev/com.ibm.ws.transport.http/src/com/ibm/ws/genericbnf/PasswordNullifier.java]: test

File: dev/com.ibm.ws.transport.http/src/com/ibm/ws/genericbnf/PasswordNullifier.java [hunk 3]
```diff
@@ -59,6 +72,36 @@ public static String nullify(String value, byte delimiter) {
...
+    public static String nullify(String header, String value, byte delimiter) {
+        // check to see if we need to null out passwords
+        if (null == value) {
+            return null;
+        }
+
+        String source = value.toLowerCase();
+        StringBuilder b = new StringBuilder(value);
+        boolean modified = optionallyMaskChars(b, source, delimiter, PASSWORD_PATTERN);
+        modified = optionallyMaskChars(b, source, delimiter, CLIENT_SECRET_PATTERN) || modified;
```
Comment (una-tapa) [inline dev/com.ibm.ws.transport.http/src/com/ibm/ws/genericbnf/PasswordNullifier.java]: @wtlucy  Thanks, this looks more professional.

File: dev/com.ibm.ws.transport.http/src/com/ibm/ws/genericbnf/PasswordNullifier.java (hunk 2 above)
Comment (una-tapa) [inline dev/com.ibm.ws.transport.http/src/com/ibm/ws/genericbnf/PasswordNullifier.java]: here?  I am clicking weird place.  Sorry @wtlucy  if many notifications were sent.   Hope this looks better.

**Comment by una-tapa** [comment]
#build

**Comment by una-tapa** [comment]
Oops, I was going to submit build first before adding reviewer (Bill,Bill,Manuel)   I will ask again via Slack when it is done.  Thanks!

File: dev/com.ibm.ws.transport.http/src/com/ibm/ws/genericbnf/PasswordNullifier.java [hunk 4]
```diff
@@ -19,25 +19,38 @@
     /** Search patterns to use for passwords */
     private static final String PASSWORD_PATTERN = "password=";
     private static final String CLIENT_SECRET_PATTERN = "client_secret=";
+    private static final String BASICAUTH_PATTERN = "basic ";
```
Comment (mrsaldana) [inline dev/com.ibm.ws.transport.http/src/com/ibm/ws/genericbnf/PasswordNullifier.java]: Should this be lower cased? If the header value contains "Basic " and assuming the source.indexOf(pattern) invocation at line 107 is case-sensitive, wouldn't that be problematic?

**Comment by una-tapa** [comment]
I wonder if I lost @mrsaldana's review.. 
Manuel
Thanks, for looking.  I originally tried "Basic" that did not work.  Then I saw this ```public static String nullify(String header, String value, byte delimiter) {
        // check to see if we need to null out passwords
        if (null == value) {
            return null;
        }

        String source = value.toLowerCase();```  So it had to be lowercase to get nullify to work..

File: dev/com.ibm.ws.webcontainer.security/src/com/ibm/ws/webcontainer/security/WebAppSecurityCollaboratorImpl.java [hunk 5]
```diff
@@ -587,6 +603,77 @@ public Object preInvoke(HttpServletRequest req, HttpServletResponse resp, String
...
+            sb.append(DEBUG_CONTEXT_PATH).append(req.getContextPath());
+        if (req.getServletPath() != null)
+            sb.append(DEBUG_SERVLET_PATH).append(req.getServletPath());
+        if (req.getPathInfo() != null)
+            sb.append(DEBUG_PATH_INFO).append(req.getPathInfo());
+        return sb.toString();
+    }
+
+    /**
+     * @r5
```
Comment (mrsaldana) [inline dev/com.ibm.ws.webcontainer.security/src/com/ibm/ws/webcontainer/security/WebAppSecurityCollaboratorImpl.java]: Should the documentation on this method description be re-evaluated?

Comment (una-tapa) [reply dev/com.ibm.ws.webcontainer.security/src/com/ibm/ws/webcontainer/security/WebAppSecurityCollaboratorImpl.java]: Yes,  I will update it. That's good comment.  Now I realize  I copy and pasted tWAS method.   I will work on it.

File: dev/com.ibm.ws.webcontainer.security/src/com/ibm/ws/webcontainer/security/WebAppSecurityCollaboratorImpl.java [hunk 6]
```diff
@@ -587,6 +603,78 @@ public Object preInvoke(HttpServletRequest req, HttpServletResponse resp, String
...
+        if (req.getContextPath() != null)
+            sb.append(DEBUG_CONTEXT_PATH).append(req.getContextPath());
+        if (req.getServletPath() != null)
+            sb.append(DEBUG_SERVLET_PATH).append(req.getServletPath());
+        if (req.getPathInfo() != null)
+            sb.append(DEBUG_PATH_INFO).append(req.getPathInfo());
+        return sb.toString();
+    }
+
+    /**
```
Comment (una-tapa) [inline dev/com.ibm.ws.webcontainer.security/src/com/ibm/ws/webcontainer/security/WebAppSecurityCollaboratorImpl.java]: Hope this header looks better.

**Comment by una-tapa** [comment]
#build

**Comment by una-tapa** [comment]
Checked the latest build after incorporating all the suggestions.   
- Unit Test looks good.  
- FVT has 2 errors but they  have been there, not related to this code.  
- Manual test looks good as well 
I am going to ask for review on Monday! 
``` 
output 
[10/27/19 10:52:08:383 EDT] 0000005c id=00000000 com.ibm.ws.webcontainer.srt.SRTServletRequest .. name --> Authorization header --> Basic ****************
<[10/27/19 10:52:08:384 EDT] 0000005c id=00000000 .ibm.ws.webcontainer.security.WebAppSecurityCollaboratorImpl 3 Http Header names and values:
Host=[localhost:9080]
User-Agent=[Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:70.0) Gecko/20100101 Firefox/70.0]
Accept=[text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8]
Accept-Language=[en-US,en;q=0.5]
Accept-Encoding=[gzip, deflate]
Authorization=[Basic xxxxx]
Connection=[keep-alive]
Cookie=[LtpaToken2=sS3aPrh3l....IcPxpJ3LrcTKDKvn9gTyoFSRc5Nom92mJ0TEh]
Upgrade-Insecure-Requests=[1]
Cache-Control=[max-age=0]
Request Context Path=/basicauth, Servlet Path=/servlet
```

**Comment by una-tapa** [comment]
Oh, my goodness.  Thank you so much!  Merging.

**Pull request merged**